File: app/src/main/java/com/boardgamegeek/sorter/LastViewedSorter.java
package com.boardgamegeek.sorter;

import android.content.Context;
import android.database.Cursor;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.text.format.DateUtils;

import com.boardgamegeek.R;
import com.boardgamegeek.provider.BggContract.Games;

public class LastViewedSorter extends CollectionSorter {
	@NonNull private final String never;

	public LastViewedSorter(@NonNull Context context) {
		super(context);
		descriptionId = R.string.menu_collection_sort_last_viewed;
		orderByClause = getClause(Games.LAST_VIEWED, true);
		never = context.getString(R.string.never);
	}

	@StringRes
	@Override
	public int getTypeResource() {
		return R.string.collection_sort_type_last_viewed;
	}

	@Override
	public String[] getColumns() {
		return new String[] { Games.LAST_VIEWED };
	}

	@NonNull
	@Override
	public String getHeaderText(@NonNull Cursor cursor) {
		long time = getLong(cursor, Games.LAST_VIEWED);
		if (time == 0) {
			return never;
		}
		return DateUtils.getRelativeTimeSpanString(time, System.currentTimeMillis(), DateUtils.DAY_IN_MILLIS).toString();
	}

	@NonNull
	@Override
	public String getDisplayInfo(@NonNull Cursor cursor) {
		long time = getLong(cursor, Games.LAST_VIEWED);
		if (time == 0) {
			return never;
		}
		return DateUtils.getRelativeTimeSpanString(time, System.currentTimeMillis(), DateUtils.MINUTE_IN_MILLIS).toString();
	}
}

File: app/src/main/java/com/boardgamegeek/sorter/LastViewedSorter.kt
package com.boardgamegeek.sorter

import android.content.Context
import androidx.annotation.StringRes
import com.boardgamegeek.R
import com.boardgamegeek.entities.CollectionItem
import com.boardgamegeek.extensions.asPastDaySpan

class LastViewedSorter(context: Context) : CollectionSorter(context) {
    override val ascendingSortTypeResId: Int
        @StringRes
        get() = R.string.collection_sort_type_last_viewed_asc

    override val descendingSortTypeResId: Int
        @StringRes
        get() = R.string.collection_sort_type_last_viewed

    override val descriptionResId: Int
        @StringRes
        get() = R.string.collection_sort_last_viewed

    override fun sortAscending(items: Iterable<CollectionItem>) = items.sortedBy { it.lastViewedDate }

    override fun sortDescending(items: Iterable<CollectionItem>) = items.sortedByDescending { it.lastViewedDate }

    override fun getHeaderText(item: CollectionItem) = item.lastViewedDate.asPastDaySpan(context).toString()

    override fun getDisplayInfo(item: CollectionItem) = ""

    override fun getTimestamp(item: CollectionItem) = item.lastViewedDate
}
